partiql-eval/src/eval.rs

```diff
@@ -46,56 +44,172 @@ pub enum EvaluationError {
+impl Evaluable for EvalScan {
     fn evaluate(&mut self, ctx: &dyn EvalContext) -> Option<Value> {
-        let mut value = partiql_bag!();
+        let mut value = partiql_bag![];
         let v = self.expr.evaluate(&Tuple(HashMap::new()), ctx);
+        let ordered = &v.is_ordered();
+        let mut c: i64 = 0;
+        let at_key = &self.at_key;
         for t in v.into_iter() {
-            let out = Tuple(HashMap::from([(self.as_key.clone(), t)]));
-            value.push(Value::Tuple(Box::new(out)));
+            let mut out = HashMap::from([(self.as_key.clone(), t)]);
+            if !at_key.is_empty() {
+                let at_id = if *ordered { c.into() } else { Missing };
+                out.insert(at_key.clone(), at_id.into());
+                c += 1;
+            }
+
+            value.push(Value::Tuple(Box::new(Tuple(out))));
         }
-        Some(Value::Bag(Box::new(value)))
+
+        self.output = Some(Value::Bag(Box::new(value)));
+        self.output.clone()
     }
+
     fn update_input(&mut self, _input: &Value) {
         todo!("update_input for Scan")
     }
 }
 
 #[derive(Debug)]
-pub struct Project {
+pub struct EvalUnpivot {
+    pub expr: Box<dyn EvalExpr>,
+    pub as_key: String,
+    pub at_key: String,
+    pub input: Option<Value>,
+    pub output: Option<Value>,
+}
+
+impl EvalUnpivot {
+    pub fn new(expr: Box<dyn EvalExpr>, as_key: &str, at_key: &str) -> Self {
+        EvalUnpivot {
+            expr,
+            as_key: as_key.to_string(),
+            at_key: at_key.to_string(),
+            input: None,
+            output: None,
+        }
+    }
+}
+
+impl Evaluable for EvalUnpivot {
+    fn evaluate(&mut self, ctx: &dyn EvalContext) -> Option<Value> {
+        let result = self.expr.evaluate(&Tuple(HashMap::new()), ctx);
+        let mut out = vec![];
+
+        let tuple = match result {
+            Value::Tuple(tuple) => *tuple,
+            other => other.coerce_to_tuple(),
+        };
+
+        let unpivoted = tuple.0.into_iter().map(|(k, v)| {
+            Tuple::from([(self.as_key.as_str(), v), (self.at_key.as_str(), k.into())])
+        });
+
+        for t in unpivoted {
+            out.push(Value::Tuple(Box::new(t)));
+        }
+
+        self.output = Some(Value::Bag(Box::new(Bag::from(out))));
+        self.output.clone()
+    }
+
+    fn update_input(&mut self, input: &Value) {
+        self.input = Some(input.clone());
+    }
+}
+
+#[derive(Debug)]
+pub struct EvalFilter {
+    pub expr: Box<dyn EvalExpr>,
+    pub input: Option<Value>,
+    pub output: Option<Value>,
+}
+
+impl EvalFilter {
+    pub fn new(expr: Box<dyn EvalExpr>) -> Self {
+        EvalFilter {
+            expr,
+            input: None,
+            output: None,
+        }
+    }
+
+    #[inline]
+    fn eval_filter(&self, bindings: &Tuple, ctx: &dyn EvalContext) -> bool {
+        let result = self.expr.evaluate(bindings, ctx);
+        match result {
+            Boolean(bool_val) => bool_val,
```

Comment:
Could we also have a `NULL` or `MISSING` value? E.g. `WHERE` expression evaluates to `NULL` or `MISSING`

Reply:
Good catch, I think spec. specifically discusses about this case; let me look into it.